### PR TITLE
Generate autopilot controller

### PR DESCRIPTION
**Pull request by ilackarms**
generates a per-resource controller which watches generic k8s events

### CODE REVIEW

File: codegen/render/kube_types_renderer.go [hunk 1]
```diff
@@ -3,82 +3,118 @@ package render
...
+	},
+	"code/types/register.gotmpl": {
+		Path: "register.go",
+	},
+	"code/types/doc.gotmpl": {
+		Path: "doc.go",
+	},
+	"code/types/proto_deepcopy.gotmpl": {
+		Path: "proto_deepcopy.go",
+	},
```
Comment (ilackarms) [inline codegen/render/kube_types_renderer.go:55]: this file should only be generated when the top level `RenderProtos == true`

File: codegen/model/template_data.go [hunk 2]
```diff
@@ -52,7 +51,8 @@ type ProjectData struct {
 }
 
 func NewTemplateData(project v1.AutopilotProject, operator v1.AutopilotOperator, templates packr.Box) (*ProjectData, error) {
-	projectGoPkg := util.GetGoPkg()
+	//projectGoPkg := util.GetGoPkg()
+	projectGoPkg := "not implemented"
```
Comment (EItanya) [inline codegen/model/template_data.go:55]: This function is used other places, why remove it here?

File: codegen/render/render_test.go [hunk 3]
```diff
@@ -11,13 +11,19 @@ import (
 	"k8s.io/apimachinery/pkg/runtime/schema"
 )
 
-var _ = Describe("Render", func() {
+// this test deprecated in favor of codegen/cmd_test
```
Comment (EItanya) [inline codegen/render/render_test.go]: can we just remove it then?

File: codegen/writer/writer.go [hunk 4]
```diff
@@ -10,29 +10,20 @@ import (
 	"strings"
 )
 
-type Writer struct {
-	Root           string
-	ForceOverwrite bool
+type Writer interface {
```
Comment (EItanya) [inline codegen/writer/writer.go]: Can we name this something else? Writer is a very well established interface

File: pkg/controller/watch_manager.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+//	Watch(res metav1.Object, h handler.EventHandler) (context.CancelFunc, error)
+//}
+//
+//type watcher struct {
+//	c controller.Controller
+//}
+//
+//func (w *watcher) Watch(res metav1.Object, h handler.EventHandler) (context.CancelFunc, error) {
+//	w.c.Watch()
+//}
```
Comment (EItanya) [inline pkg/controller/watch_manager.go]: Delete file

File: test/e2e/run_test_project.sh [hunk 6]
```diff
@@ -39,9 +39,12 @@ function generate_traffic {
 }
 
 function cleanup {
+    set -x
     kubectl delete ns canary-operator
     kubectl delete ns e2etest
     kubectl delete crd canarydeployments.autopilot.examples.io
+    set +x
+    echo cleaned up!
 }
```
Comment (EItanya) [inline test/e2e/run_test_project.sh:48]: Is this used?

File: cloudbuild.yaml [hunk 7]
```diff
@@ -17,11 +17,14 @@ steps:
...
     args: ['go', 'build', './...']
     waitFor: ['untar-mod-cache']
+    env:
+      - 'GO111MODULE=on'
     id: 'build'
 
   - name: 'gcr.io/$PROJECT_ID/go-mod-ginkgo:0.1.5'
     volumes: *vol
     env:
+      - 'GO111MODULE=on'
```
Comment (EItanya) [inline cloudbuild.yaml]: Formatting is off here, also you can update the version of the container to `0.2.1` for latest go version

File: test/e2e/run_test_project.sh (hunk 6 above)
Comment (ilackarms) [reply test/e2e/run_test_project.sh:48]: not currently

File: codegen/model/template_data.go (hunk 2 above)
Comment (ilackarms) [reply codegen/model/template_data.go:55]: this is a deprecated function, will be removing in a follow up

**Review by EItanya — APPROVED** [review]

**Pull request merged**